### PR TITLE
Add API support for time-varying scalars (#364)

### LINKED ISSUE #364: Add API to support time-varying scalars
From discussion with @radka-j @qiencai - currently the `Sample` has the following fields:
```python
class Sample:
    """A batch in input data space."""

    input_fields: TensorTSC
    output_fields: TensorTSC
    constant_scalars: TensorC | None
    constant_fields: TensorSC | None
    boundary_conditions: TensorS | None
```

Alongside these, it would also be useful to support:
```python
time_varying_scalars: TensorTC | None
```
i.e. a vector of scalars that vary by timestep can be passed.

With this, adding API to support passing this data as optional additional modulation for downstream processors both during training and rollout would be useful so that the processor can condition on this.

`constant_scalars` are currently passed to processors by either:
1. First encoding (such as `PermuteConcat`)
2. Passing with the`global_cond` argument

It might be simplest to support with 2. via passing to `global_cond` through concat of the slice of the `time_varying_scalars`.

Areas to update to support:
- `SpatioTemporalDataset`: in `__init__` the full series is probably including (like `constant_scalars` and `constant_fields`) - in `__getitem__` the full time series can then be included in batches
- Encoding: `def encode_with_cond()` - this could be updated to concat directly there
- Caching and processor: if handled with `global_cond` and `encode_with_cond()` there might not be any need for changes here.
- Rollout: since the rollout is performed autoregressively, the `_advance_batch()` API might need revisiting for setting which slice of time-varying scalars to make available for each prediction. There might also be better approaches to updating this method.

### PR DESCRIPTION
**Pull request by qiencai**
Adds a time-varying scalars (TVS) channel, end-to-end from datasets through the encoder's global conditioning vector and is correctly sliced during autoregressive rollouts.

## Usage
Provide a `time_varying_scalars` array of shape `(n_trajectories, n_timesteps, n_channels)` in your HDF5 data file alongside the existing data arrays, the dictionary key should be "time_varying_scalars", and set `n_tvs_extra_steps` on the datamodule to at least `autoregressive_train_steps * stride` so rollout always has a fresh slice; the encoder will then automatically concatenate the current TVS slice onto `global_cond` at every step, no model-side changes required.

## Changes
- types: new TensorTC alias; Sample/Batch carry optional time_varying_scalars with collation, repeat(), and to() support.
- data: HDF5Dataset loads per-trajectory time_varying_scalars and stores a window of n_steps_output + n_tvs_extra_steps slices per sample(which I set to 365, this is just to slice long enough TVS so the sequence won't be all consumed during rollout, such parameter needs to exist, to make sure the shape is consistent for all samples); new n_tvs_extra_steps dataset/datamodule parameter so rollout has enough future slices.
- encoders.base: encode_cond appends the current TVS slice to global_cond and raises a clear error if the buffer is exhausted.
- models.encoder_processor_decoder: _clone_batch clones TVS; _advance_batch shifts TVS by stride at every rollout step.
- scripts/setup: expose n_time_varying_scalars in logic_stats.
- scripts/data, benchmarking/inference: TVS-aware device move and synthetic batch construction.

Closes #364.

### CODE REVIEW

**Comment by qiencai** [comment]
I have the time varying fields code change locally, will follow up with that PR after we agree on implementations for the scalar case.

File: src/autocast/data/dataset.py [hunk 1]
```diff
@@ -86,6 +88,12 @@ def __init__(  # noqa: PLR0915
             Path to normalization statistics file (yaml). Defaults to None.
         normalization_stats: dict | None
             Preloaded normalization statistics. Defaults to None.
+        n_tvs_extra_steps: int
```
Comment (sgreenbury) [inline src/autocast/data/dataset.py]: Is there a specific motivation for including the extra steps here? Perhaps it can be removed since updating the sequence is handled in `_advance_batch`

File: src/autocast/benchmarking/inference.py [hunk 2]
```diff
@@ -34,6 +34,9 @@ def make_synthetic_batch(example_batch: Batch, batch_size: int) -> Batch:
         constant_scalars=_slice_optional_first(example_batch.constant_scalars),
         constant_fields=_slice_optional_first(example_batch.constant_fields),
         boundary_conditions=_slice_optional_first(example_batch.boundary_conditions),
+        time_varying_scalars=_slice_optional_first(
+            getattr(example_batch, "time_varying_scalars", None)
+        ),
     )
```
Comment (sgreenbury) [inline src/autocast/benchmarking/inference.py]: I think it would be good to retrieve in the same was as the other fields:
```suggestion
        time_varying_scalars=_slice_optional_first(example_batch.time_varying_scalars),
)
```

File: src/autocast/encoders/base.py [hunk 3]
```diff
@@ -152,6 +156,21 @@ def encode_cond(self, batch: Batch) -> Tensor | None:
...
+            if batch.time_varying_scalars.shape[1] == 0:
+                msg = (
+                    "time_varying_scalars is exhausted (shape[1] == 0). "
+                    "The autoregressive rollout has consumed all pre-computed "
+                    "steps. Increase `n_tvs_extra_steps` in the datamodule "
+                    "config so the dataset stores enough future steps."
+                )
+                raise RuntimeError(msg)
+            # Slice the current prediction step (index 0 along time axis).
+            tvs = batch.time_varying_scalars[:, 0, :]  # (B, C_tvs)
```
Comment (sgreenbury) [inline src/autocast/encoders/base.py]: Perhaps worth keeping variables with `time_varying_scalars` instead of `tvs` for clarity.

**Review by sgreenbury — COMMENTED** [review]
Thank you @qiencai for taking a look at this and opening the PR! I added some initial comments below, and the following thoughts to discuss:
- When dataset is not in full trajectory mode, would it make sense to get slices of `time_varying_scalars` that match the input temporal dim's size?
- For full trajectory model, should this provide the full sequence with inputs or on the outputs?
- Remaining aspects include incorporating into the encoder class/latent spaces

File: src/autocast/data/dataset.py [hunk 4]
```diff
@@ -186,6 +196,21 @@ def __init__(  # noqa: PLR0915
...
+                if self.time_varying_scalars is not None:
+                    t_out_start = sub_idx * self.stride + self.n_steps_input
+                    t_out_end = (
+                        t_out_start + self.n_steps_output + self.n_tvs_extra_steps
+                    )
+                    self.all_time_varying_scalars.append(
+                        self.time_varying_scalars[
+                            traj_idx, t_out_start:t_out_end, :
+                        ].to(self.dtype)
+                    )
```
Comment (sgreenbury) [inline src/autocast/data/dataset.py]: As discussed, if including the time varying scalars for the whole input window, potentially:
```suggestion
                    t_start = sub_idx * self.stride
                    t_end = sub_idx * self.stride + self.n_steps_input
                    self.all_time_varying_scalars.append(
                        self.time_varying_scalars[
                            traj_idx, t_start:t_end, :
                        ].to(self.dtype)
                    )
```

File: src/autocast/encoders/base.py (hunk 3 above)
Comment (sgreenbury) [inline src/autocast/encoders/base.py]: As discussed, could permute this instead so time and channels from across the input window are concatenated:
```suggestion
            tvs = rearrange(batch.time_varying_scalars, "b t c -> b (t c)")
```

Comment (sgreenbury) [reply src/autocast/encoders/base.py]: This is also potentially where the length of the window could be adjusted (e.g. a parameter on the encoder like e.g. `n_time_varying_scalars=1`:
```python
n_steps_input = batch.input_fields.shape[1]
tvs = rearrange(batch.time_varying_scalars[:, n_steps_input-n_time_varying_scalars:n_steps_input, :] , "b t c -> b (t c)")
```

**Review by sgreenbury — COMMENTED** [review]
Thanks for the discussion earlier, adding some notes/code here on the ideas:

- Update dataset windowing: we could aim to have the `n_steps_input` window for the time varying scalars (so same length of time as input as the `input_fields`.
- Permuting the scalars in encode
- Add flexibility around `n_steps_input` being different to `n_time_varying_scalars`
- Tests: it could work well to add some tests using the `time_varying_scalars` like some of the examples in this test file https://github.com/alan-turing-institute/autocast/blob/3cf33ddcfa80ab78375212809824b9c06595621c/tests/models/test_encoder_processor_decoder.py

File: src/autocast/data/dataset.py (hunk 4 above)
Comment (sgreenbury) [reply src/autocast/data/dataset.py]: Just to add that if in full trajectory mode (as used in the rollout) I think this would need to also include the output steps within the input here since the rollout would need to be able to stride the time dim.
```python
t_end = sub_idx * self.stride + self.n_steps_input
t_end = t_end + self.n_steps_output if full_trajectory_mode else t_end
time_varying_scalars = self.time_varying_scalars[traj_idx, t_start : t_start + t_end]
self.all_time_varying_scalars.append(time_varying_scalars)
```

Comment (qiencai) [reply src/autocast/data/dataset.py]: I think you are right. I was confused because I moved away from `full_trajectory_mode` after our last discussion.

I needed to **initialize forecasts from any date inside a trajectory, and rollout for a prescribed lead time**, so I now use a windowed rollout dataloader. It constructs samples as follows:

Given a randomly generated index `i`, dataset `X`, and `t_start = i * stride`:

* input window: `X[t_start : t_start + n_steps_input]`
* truth: `X[t_start + n_steps_input : t_start + n_steps_input + max_rollout_steps * stride]`

Here, the truth is the future states we would ever need, depending on how long we want to roll out for.

This is also more memory-friendly. In fact, I moved to this because of an OOM issue when I loaded my batch with a long sequence of time-varying fields.

I think for full traj mode we should stick to what you've suggested, and add extra conditioning for the windowed data loader.....

Comment (sgreenbury) [reply src/autocast/data/dataset.py]: Thanks @qiencai for the comment above and discussion earlier, adding the code we sketched to convert a single trajectory into subtrajectories for rollouts:
```python
def get_indices_of_month_starts() -> list[int]:
    """Get the indices of the starts of each month in the dataset."""
    return list(range(0, 100, 30))  # Example: months at indices 0, 30, 60, etc.


def convert_to_subtrajectory(
    x: torch.Tensor, start_idxs: list[int], traj_idx: int, window_size: int
) -> torch.Tensor:
    """Convert a tensor into subtajectories based on provided start indices.

    x: Tensor of shape (N, T, ...)
    traj_idx: Which trajectory to convert to subtrajectories
    start_idxs: List of starting indices for each chunk
    window_size: Size of each subtrajectory window (number of time steps)
    """
    return torch.stack(
        [
            x[
                traj_idx,
                start_idx : start_idx + window_size,
            ]
            for start_idx in start_idxs
        ],
        dim=0,
    )
```

It might be worth us trying to add this functionality directly on to the `SpatiotemporalDataset` too? Maybe a new mode (`subtrajectory_mode` to go alongside `full_trajectory_mode`?) that could handle this directly and could be used to optionally create the rollout dataloaders instead of the full trajectory mode.

File: src/autocast/data/dataset.py [hunk 5]
```diff
@@ -47,6 +48,8 @@ def __init__(  # noqa: PLR0915
         normalization_type: type[ZScoreNormalization] | None = ZScoreNormalization,
         normalization_path: str | None = None,
         normalization_stats: dict | DictConfig | None = None,
+        subtrajectory_mode: bool = False,
+        subtrajectory_start_idxs: list[int] | None = None,
```
Comment (sgreenbury) [inline src/autocast/data/dataset.py:52]: Looking at the design again, I think some additional API is needed in order to enable the rollout for each subtrajectory to have sufficient length for the rollout?
```suggestion
        subtrajectory_start_idxs: list[int] | None = None,
        n_steps_rollout: int | None = None
```

File: src/autocast/data/dataset.py [hunk 6]
```diff
@@ -146,16 +178,14 @@ def __init__(  # noqa: PLR0915
         self.all_output_fields = []
         self.all_constant_scalars = []
         self.all_constant_fields = []
+        self.all_time_varying_scalars: list[torch.Tensor] = []
+
+        # Each sample spans `n_steps_input + n_steps_output` steps.
+        window_size = self.n_steps_input + self.n_steps_output
```
Comment (sgreenbury) [inline src/autocast/data/dataset.py:184]: With the above suggestion re. `n_steps_rollout`, I think this would be updated to?
```suggestion
        window_size = self.n_steps_input + self.n_steps_rollout
```

File: src/autocast/data/dataset.py [hunk 7]
```diff
@@ -86,6 +89,20 @@ def __init__(  # noqa: PLR0915
...
+        subtrajectory_mode: bool
+            If True, build rollout subtrajectories at explicit start indices
+            (`subtrajectory_start_idxs`) instead of sliding a window at every
+            stride. Each subtrajectory spans `n_steps_input + n_steps_output`
+            steps; `n_steps_output` is the rollout horizon and is *not*
+            auto-expanded to the full trajectory (unlike `full_trajectory_mode`).
+            Use `data.calendar.month_start_indices` to generate month-anchored
+            starts. Cannot be combined with `full_trajectory_mode` or
+            `autoencoder_mode`. Defaults to False.
+        subtrajectory_start_idxs: list[int] | None
```
Comment (sgreenbury) [inline src/autocast/data/dataset.py:101]: I think we're assuming: `subtrajectory_start_idxs: list[int] | None = None` for each trajectory with the current API. I think this would ideally a list per trajecrory  so possibly could be extended to something like:
```
subtrajectory_start_idxs: list[int] | dict[int, list[int]] | None
            Input-window start indices for `subtrajectory_mode`. A flat list is
            applied to every trajectory. A dict maps trajectory index to the
            starts for that trajectory, and missing trajectories contribute no
            rollout subtrajectories. Required when `subtrajectory_mode` is True.
            Defaults to None.
```

File: src/autocast/data/calendar.py
Comment (sgreenbury) [inline src/autocast/data/calendar.py:1]: If we could document and update this to be documented for any time series data where days are the unit, thaat would be great.

**Review by sgreenbury — COMMENTED** [review]
Thanks for the updates @qiencai, I think this is shaping up really well. Added some initial comments below for discussion.

File: src/autocast/data/calendar.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        end = date(year, 12, 31)
+        while d <= end:
+            is_leap_day = d.month == 2 and d.day == 29
+            if not (drop_leap_day and is_leap_day and calendar.isleap(d.year)):
+                dates.append(d)
+            d += timedelta(days=1)
+    return dates
+
+
+def month_start_indices(
```
Comment (sgreenbury) [inline src/autocast/data/calendar.py:45]: From discussion with @qiencai, adding a comment here around the idea of enabling this to be potentially more hydra-configurable with something like:
- An abstract base class to describe index generators (`class SubtrajectoryIndexGenerator(abc.ABC)`)
- Some subclasses (e.g. `class MonthStartIndexGenerator(SubtrajectoryIndexGenerator)`

Sketch below that could go in a separate module in e.g. `src/autocast/data/subtrajectory.py`
```python

# i.e. list or dict of lists but for hydra
SubtrajectoryStartIdxs = Sequence[int] | Mapping[int | str | Sequence[int]]

class SubtrajectoryIndexGenerator(abc.ABC):
    """Generate rollout subtrajectory start indices for a dataset split."""

    @abc.abstractmethod
    def __call__(
        self,
        *,
        n_trajectories: int,
        n_timesteps: int,
        n_steps_input: int,
        n_steps_rollout: int,
    ) -> SubtrajectoryStartIdxs:
        """Return flat or per-trajectory rollout start indices."""

@dataclass
class MonthStartIndexGenerator(SubtrajectoryIndexGenerator):
    """Generate starts whose initialization date falls on the first of a month."""

    years: list[int]
    stride: int = 1
    drop_leap_day: bool = True

    def __call__(
        self,
        *,
        n_trajectories: int,
        n_timesteps: int,
        n_steps_input: int,
        n_steps_rollout: int,
    ) -> list[int]:
        """Return month-start initialization starts shared by every trajectory."""
        start_idxs, _ = month_start_indices(
            test_years=self.years,
            n_steps_input=n_steps_input,
            max_rollout_steps=n_steps_rollout, # NB. I think this might need changing in current form since max_rollout steps is rollouts not timesteps.
            stride=self.stride,
            drop_leap_day=self.drop_leap_day,
        )
        return start_idxs

```
